### PR TITLE
Initial missing numbers commit

### PR DESCRIPTION
**Pull request by SeR4h**
@dennisja please review my code. it includes all tests.
Write a javascript program/function to find missing numbers from an array. Follow aTest Driven Development approach while completing this challenge

### CODE REVIEW

File: missingNumbers.js [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+    while (missingNumbers.length < count) {
+        while (inputArray[i] !== missingValue) {
+            missingNumbers.push(missingValue++)
+        }
+        i++;
+        missingValue++;
+    }
+    return missingNumbers;
+}
+/*var ar = [ 4,9,1 ];
```
Comment (dennisja) [inline missingNumbers.js]: It's always a good practice to remove commented code. It's always a sign that you don't need it.
Comments are usually used to explain why you are doing something in code

File: missingNumbers.js [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+    if (typeof inputArray === 'string') {
+        return 'invalid input'
+    }
+    if (inputArray.length <= 1) {
+        return 'array length too small'
+    }
+    inputArray = inputArray.sort()
+    var i = 1
+    var missingValue = inputArray[0] + 1;
+    const count = ((inputArray[inputArray.length - 1]) - missingValue) - (inputArray.length - 2);
```
Comment (dennisja) [inline missingNumbers.js]: A few questions on this?
- What count doe this variable keep track of?
- Why do you do this `inputArray.length - 2`, How did you come up with 2

File: missingNumbers.js [hunk 3]
```diff
@@ -0,0 +1,25 @@
+function findMissingNumbers(inputArray) {
+
+    var missingNumbers = [];
+    if (typeof inputArray === 'string') {
+        return 'invalid input'
```
[the remark this answers is not in the document]
Comment (SeR4h) [reply missingNumbers.js]: @dennisja I tried the throw but the tests were failing so i still used 'return'

**Comment by SeR4h** [comment]
@dennisja please check the corrections i made on the function.

File: missingNumbers.js [hunk 4]
```diff
@@ -1,25 +1,23 @@
 function findMissingNumbers(inputArray) {
-
     var missingNumbers = [];
-    if (typeof inputArray === 'string') {
-        return 'invalid input'
+    if (!Array.isArray(inputArray)) {
+        // used return for testing since thrown error couldnt be seen by the test
+        return 'Invalid Input'
```
Comment (dennisja) [inline missingNumbers.js]: Throw the error and modify the tests as discussed [here ](https://github.com/SeR4h/andelaweek3challenges1/pull/2#discussion_r359166870)

File: missingNumbers.js [hunk 5]
```diff
@@ -1,25 +1,23 @@
...
     }
     if (inputArray.length <= 1) {
         return 'array length too small'
```
Comment (dennisja) [inline missingNumbers.js]: Throw the error and modify the tests as discussed [here ](https://github.com/SeR4h/andelaweek3challenges1/pull/2#discussion_r359166870)

File: missingNumbers.js [hunk 6]
```diff
@@ -1,25 +1,23 @@
...
-    var missingValue = inputArray[0] + 1;
-    const count = ((inputArray[inputArray.length - 1]) - missingValue) - (inputArray.length - 2);
-    while (missingNumbers.length < count) {
-        while (inputArray[i] !== missingValue) {
-            missingNumbers.push(missingValue++)
+    inputArray.sort((a, b) => a - b)
+    for (var i = 1; i < inputArray.length; i++) {
+        if (inputArray[i] - inputArray[i - 1] != 1) {
+            var valDifference = inputArray[i] - inputArray[i - 1];
+            var j = 1
```
Comment (dennisja) [inline missingNumbers.js]: Make this a `for loop` too

Comment (dennisja) [reply missingNumbers.js]: I meant this
```js
var j = 1
 while (j < valDifference) {
     missingNumbers.push(inputArray[i - 1] + j)
      j++
}
```

File: package.json [hunk 7]
```diff
@@ -0,0 +1,23 @@
...
+  },
+  "keywords": [],
+  "author": "",
+  "license": "ISC",
+  "bugs": {
+    "url": "https://github.com/SeR4h/andelaweek3optionalChallenge/issues"
+  },
+  "homepage": "https://github.com/SeR4h/andelaweek3optionalChallenge#readme",
+  "dependencies": {
+    "jest": "^24.9.0"
```
[the remark this answers is not in the document]
Comment (dennisja) [reply package.json]: @SeR4h please resolve this

File: missingNumbers.js (hunk 4 above)
Comment (SeR4h) [reply missingNumbers.js]: @dennisja. i have reinstalled jest as devDependence

**Comment by SeR4h** [comment]
@dennisja please review ''correction on tests and final missing numbers." i've made all corrections.

File: missingNumbers.js [hunk 8]
```diff
@@ -0,0 +1,20 @@
...
+    if (!Array.isArray(inputArray)) {
+        throw new Error('Invalid Input');
+    }
+    if (inputArray.length <= 1) {
+        throw new Error('array length too small');
+    }
+    inputArray.sort((a, b) => a - b)
+    for (var i = 1; i < inputArray.length; i++) {
+        if (inputArray[i] - inputArray[i - 1] != 1) {
+            var valDifference = inputArray[i] - inputArray[i - 1];
```
Comment (dennisja) [inline missingNumbers.js]: Move this line before the `if` statement and use the variable in the `if condition`

File: missingNumbers.js [hunk 9]
```diff
@@ -0,0 +1,20 @@
...
+    var missingNumbers = [];
+    if (!Array.isArray(inputArray)) {
+        throw new Error('Invalid Input');
+    }
+    if (inputArray.length <= 1) {
+        throw new Error('array length too small');
+    }
+    inputArray.sort((a, b) => a - b)
+    for (var i = 1; i < inputArray.length; i++) {
+        if (inputArray[i] - inputArray[i - 1] != 1) {
```
Comment (dennisja) [inline missingNumbers.js]: Change this condition from `!=` to `>`. By doing that we don't run anything in the if statement if the difference is1 or less

File: missingNumbers.js [hunk 10]
```diff
@@ -0,0 +1,20 @@
...
+            var valDifference = inputArray[i] - inputArray[i - 1];
+            for (j = 1; j < valDifference; j++) {
```
Comment (dennisja) [inline missingNumbers.js]: You have not initialized `j` anywhere

File: package.json [hunk 11]
```diff
@@ -0,0 +1,24 @@
...
+  "license": "ISC",
+  "bugs": {
+    "url": "https://github.com/SeR4h/andelaweek3optionalChallenge/issues"
+  },
+  "homepage": "https://github.com/SeR4h/andelaweek3optionalChallenge#readme",
+  "dependencies": {},
+  "devDependencies": {
+    "jest": "^24.9.0"
+  }
+}
```
Comment (dennisja) [inline package.json]: Add an end of line at the end of this file. Just autoformat the code using your text editor

**Comment by SeR4h** [comment]
@dennisja please check resolved changes. I've made all the corrections

File: missingNumbers.js [hunk 12]
```diff
@@ -0,0 +1,20 @@
...
+    if (!Array.isArray(inputArray)) {
+        throw new Error('Invalid Input');
+    }
+    if (inputArray.length <= 1) {
+        throw new Error('array length too small');
+    }
+    inputArray.sort((a, b) => a - b)
+    for (var i = 1; i < inputArray.length; i++) {
+        var valDifference = inputArray[i] - inputArray[i - 1];
+        if (valDifference > 1) {
```
Comment (dennisja) [inline missingNumbers.js]: I, by the way, think we don't need this `if condition` the for loop initialization and condition cater for it

**Pull request merged**